gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/store/ManagedDirectBufferAllocator.java

```diff
@@ -150,24 +148,29 @@ private ByteBuffer allocate(String objectName, int size,
   public ByteBuffer expand(ByteBuffer buffer, int required, String owner) {
     assert required > 0 : "expand: unexpected required = " + required;
 
-    final int currentUsed = buffer.capacity();
-    final int newLength = UnsafeHolder.getAllocationSize(
-        BufferAllocator.expandedSize(currentUsed, required));
-    final int delta = newLength - currentUsed;
-    // expect original owner to be ManagedDirectBufferAllocator
-    if (reserveMemory(owner, delta, false) ||
-        tryEvictData(owner, delta)) {
-      try {
-        return UnsafeHolder.reallocateDirectBuffer(buffer, newLength,
-            FreeBufferBase.class, freeBufferFactory(owner));
-      } catch (IllegalStateException ise) {
-        // un-reserve the delta bytes
-        CallbackFactoryProvider.getStoreCallbacks().releaseStorageMemory(
-            owner, delta, true);
-        throw ise;
+    final int currentUsed = buffer.limit();
+    if (currentUsed + required > buffer.capacity()) {
+      final int newLength = UnsafeHolder.getAllocationSize(
+          BufferAllocator.expandedSize(currentUsed, required));
+      final int delta = newLength - currentUsed;
+      // expect original owner to be ManagedDirectBufferAllocator
+      if (reserveMemory(owner, delta, false) ||
+          tryEvictData(owner, delta)) {
+        try {
+          return UnsafeHolder.reallocateDirectBuffer(buffer, newLength,
+              FreeBufferBase.class, freeBufferFactory(owner));
```

Comment:
Major performance benefit. Heap access is always aligned, but off-heap unaligned access is allowed (and will be used) particularly on intel platforms. Without alignment each long read can lead to 2 long reads, for example. For this reason the ColumnFormatValue serialization also adds a byte padding to make the seriailzation header aligned to word-size. In fact if you see the JDK code of ByteBuffer.allocateDirect it aligns to page-size, not even word-size!

The alignment does break down when reading different sizes of strings, for example, where there is no padding in the interest of not losing space for every string but for all other cases alignment is a big win overall. All our read/write code should specifically take care of that.